hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/OMConfigKeys.java

```diff
@@ -292,4 +292,8 @@ private OMConfigKeys() {
   public static final long OZONE_OM_ADMIN_PROTOCOL_WAIT_BETWEEN_RETRIES_DEFAULT
       = 1000;
 
+  public static final String OZONE_OM_MAX_PENDING_REQ_COUNT =
+      "ozone.om.max.pending.req.count";
+  public static final int OZONE_OM_MAX_PENDING_REQ_COUNT_DEFAULT = 10000;
```

Comment:
How was this default value decided? Is there any benchmark we can take a look at to see how the performance is affected by this setting?

Reply:
in fact , the object  cached in the `current queue` and `ready queue` of `ozoneDoubleBuffer` is `DoubleBufferEntry` and  it has a fixed size 24 bytes. 
```
2022-03-09 17:37:01,670 [OM StateMachine ApplyTransaction Thread - 0] INFO org.apache.hadoop.ozone.om.ratis.OzoneManagerDoubleBuffer: DoubleBufferEntry size: org.apache.hadoop.ozone.om.ratis.helpers.DoubleBufferEntry object internals:
 OFFSET  SIZE                                                   TYPE DESCRIPTION                               VALUE
      0     4                                                        (object header)                           05 00 00 00 (00000101 00000000 00000000 00000000) (5)
      4     4                                                        (object header)                           00 00 00 00 (00000000 00000000 00000000 00000000) (0)
      8     4                                                        (object header)                           58 b9 9c 02 (01011000 10111001 10011100 00000010) (43825496)
     12     4   org.apache.hadoop.ozone.om.response.OMClientResponse DoubleBufferEntry.response                (object)
     16     8                                                   long DoubleBufferEntry.trxLogIndex             25
Instance size: 24 bytes\
```
so , 10000 DoubleBufferEntry will take about 256MB memory, it is small enough for almost all the physical machine.

i have test the size from 10000 to 300000, and i find the performance almost does not change. what is more, i find a case that  the size of double buffer flush batch is 200000, and the flush option take about 10 senconds. thus i think this value should not be too big and i think 10000 maybe a suitable value. do you have any suggesstion about this?

Reply:
> in fact , the object cached in the current queue and ready queue of ozoneDoubleBuffer is DoubleBufferEntry and it has a fixed size 24 bytes.

DoubleBufferEntry has the transaction log index and the client response. The client response size would change based on the request type. For example, KeyCreate Request would have a bigger response size than 24 bytes while SetBucketProperty would have a lesser size. But the size should not be too big irrespective of this.

> so , 10000 DoubleBufferEntry will take about 256MB memory, it is small enough for almost all the physical machine.
Taking 24 bytes, 10,000 DoubleBufferEntries would take by only 256KB of memory. 

Even if we take the average DoubleBufferEntry size as 1KB, 200000 entries should not take up around 200MB of space. 
So I am not sure why the DoubleBuffer is taking up so much space.
Could you please share your debug analysis to help me understand why DoubleBuffer is taking up too much memory?

Reply:
i am sorry i can not find those dump file for now, i forget where i put them. but  it is not hard to reproduce this situation. when benchmarking , we started 30 freon instances (in different machine), each will start 10 clients thread to write object.  
after i add the patch to our production cluster,  it works well under a heavy current pressure now, and the performance keeps the same as before.

double buffer has two unlimited buffer: current buffer and ready buffer . in fact ,  if we do not limit the buffer size, several problems comes。
1 a big ready buffer will produce a big rocksdb write batch.  for rocksdb , if a batch is too big , the write performance will downgrade。meanwhile ，a big write batch will also take much space.
2 if rocksdb spend too much time on commiting a big write batch, then more and more request will be put in the current buffer of DoubleBuffer. a big current buffer will also take much space and  it will change to a big ready buffer and then a big write batch. It will be a vicious circle。what is more, if a full GC happens, The situation will get worse